lib/Dialect/TritonGPU/Transforms/Utility.cpp

```diff
@@ -9,47 +9,91 @@
 
 namespace mlir {
 
-SmallVector<unsigned, 3> mmaVersionToInstrShape(int version,
-                                                const ArrayRef<int64_t> &shape,
-                                                RankedTensorType type) {
+static unsigned getValidNOfMmaV3(const int64_t N, Type eltType) {
+  SmallVector<unsigned> validN;
+
+  // MMAv3 with larger instruction shape is preferred.
+  if (eltType.isFloat8E5M2() || eltType.isFloat8E4M3FNUZ() || eltType.isF16() ||
+      eltType.isBF16() || eltType.isF32()) {
+    validN.assign({256, 248, 240, 232, 224, 216, 208, 200, 192, 184, 176,
+                   168, 160, 152, 144, 136, 128, 120, 112, 104, 96,  88,
+                   80,  72,  64,  56,  48,  40,  32,  24,  16,  8});
+  }
+
+  if (eltType.isInteger(8)) {
+    validN.assign({224, 208, 192, 176, 160, 144, 128, 112, 96, 80, 64, 48, 32,
+                   24, 16, 8});
+  }
+
+  for (auto n : validN) {
+    if (N % n == 0) {
+      return n;
+    }
+  }
+  return 0;
+}
+
+static SmallVector<unsigned, 3>
+mmaVersionToInstrShapeImpl(int version, const ArrayRef<int64_t> &shape,
+                           Type eltType) {
   if (version == 1)
     return {16, 16};
   else if (version == 2)
     return {16, 8};
   else if (version == 3) {
-    unsigned k = 256 / type.getElementTypeBitWidth();
+    unsigned k = 256 / eltType.getIntOrFloatBitWidth();
+    unsigned n = getValidNOfMmaV3(shape[1], eltType);
```

Comment:
should `shape[1]` also be part of the layout somehow? It seems like that the only reason we are passing those shapes to all the functions. Is there a reasonable way to keep the layout self contained?

Reply:
If shape[1] is part of the layout, this will block mma_v3 with shape(64x128x16) fuse with another mma_v3 of shape(64x64x16).

Reply:
how does it work if the same layout encoding may have different instruction shape? Doesn't it mean there is a potential mismatch?

Reply:
because first gemm acc output's N dimension is treated as K dimension for second gemm., so the two gemm's N dimension not necessary be equal.